src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -360,56 +615,62 @@ void CoreWorkerDirectTaskSubmitter::PushNormalTask(
+
+          if (reply.worker_exiting()) {
+            RAY_LOG(DEBUG) << "Worker " << addr.worker_id
+                           << " replied that it is exiting.";
+            // The worker is draining and will shutdown after it is done. Don't return
+            // it to the Raylet since that will kill it early.
+            worker_to_lease_entry_.erase(addr);
+            auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+            scheduling_key_entry.active_workers.erase(addr);
+            if (scheduling_key_entry.CanDelete()) {
+              // We can safely remove the entry keyed by scheduling_key from the
+              // scheduling_key_entries_ hashmap.
+              scheduling_key_entries_.erase(scheduling_key);
+            }
+          } else if (!reply.task_stolen() && (!status.ok() || !is_actor_creation)) {
+            // Successful actor creation leases the worker indefinitely from the raylet.
+            OnWorkerIdle(addr, scheduling_key,
+                         /*error=*/!status.ok(), assigned_resources);
+          }
 
-      // Decrement the total number of tasks in flight to any worker with the current
-      // scheduling_key.
-      auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
-      RAY_CHECK(scheduling_key_entry.active_workers.size() >= 1);
-      RAY_CHECK(scheduling_key_entry.total_tasks_in_flight >= 1);
-      scheduling_key_entry.total_tasks_in_flight--;
-    }
-    if (reply.worker_exiting()) {
-      // The worker is draining and will shutdown after it is done. Don't return
-      // it to the Raylet since that will kill it early.
-      absl::MutexLock lock(&mu_);
-      worker_to_lease_entry_.erase(addr);
-      auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
-      scheduling_key_entry.active_workers.erase(addr);
-      if (scheduling_key_entry.CanDelete()) {
-        // We can safely remove the entry keyed by scheduling_key from the
-        // scheduling_key_entries_ hashmap.
-        scheduling_key_entries_.erase(scheduling_key);
-      }
-    } else if (!status.ok() || !is_actor_creation) {
-      // Successful actor creation leases the worker indefinitely from the raylet.
-      absl::MutexLock lock(&mu_);
-      OnWorkerIdle(addr, scheduling_key,
-                   /*error=*/!status.ok(), assigned_resources);
-    }
-    if (!status.ok()) {
-      // TODO: It'd be nice to differentiate here between process vs node
-      // failure (e.g., by contacting the raylet). If it was a process
-      // failure, it may have been an application-level error and it may
-      // not make sense to retry the task.
-      RAY_UNUSED(task_finisher_->PendingTaskFailed(
-          task_id, is_actor ? rpc::ErrorType::ACTOR_DIED : rpc::ErrorType::WORKER_DIED,
-          &status));
-    } else {
-      task_finisher_->CompletePendingTask(task_id, reply, addr.ToProto());
-    }
-  });
+          if (reply.task_stolen()) {
```

Comment:
Nit: merge this into the above if-else block.

Can you also add a comment to explain why we don't need to do anything here?